File: crypto-native/ios/CryptoNativeModule.swift
import ExpoModulesCore
import CryptoKit
import Security

public class CryptoNativeModule: Module {
  public func definition() -> ModuleDefinition {
    Name("CryptoNative")

    // Generate a random salt
    AsyncFunction("generateSalt") { (length: Int) -> [UInt8] in
      var salt = [UInt8](repeating: 0, count: length)
      let status = SecRandomCopyBytes(kSecRandomDefault, salt.count, &salt)
      guard status == errSecSuccess else {
        throw CryptoError.failedToGenerateRandom
      }
      return salt
    }

    // Derive key using PBKDF2 (iOS doesn't have Argon2 natively, using PBKDF2-SHA256)
    AsyncFunction("deriveKey") { (password: String, salt: [UInt8], iterations: Int, keyLength: Int) -> [UInt8] in
      let passwordData = Data(password.utf8)
      let saltData = Data(salt)
      
      var derivedKey = [UInt8](repeating: 0, count: keyLength)
      let derivationStatus = CCKeyDerivationPBKDF(
        CCPBKDFAlgorithm(kCCPBKDF2),
        password,
        passwordData.count,
        salt,
        saltData.count,
        CCPseudoRandomAlgorithm(kCCPRFHmacAlgSHA256),
        UInt32(iterations),
        &derivedKey,
        keyLength
      )
      
      guard derivationStatus == kCCSuccess else {
        throw CryptoError.keyDerivationFailed
      }
      
      return derivedKey
    }

    // AES-GCM Encrypt
    AsyncFunction("encrypt") { (data: [UInt8], keyBytes: [UInt8]) -> [String: Any] in
      let plainText = Data(data)
      let key = SymmetricKey(data: Data(keyBytes))
      
      let sealedBox = try AES.GCM.seal(plainText, using: key)
      
      return [
        "ciphertext": Array(sealedBox.ciphertext),
        "nonce": Array(sealedBox.nonce),
        "tag": Array(sealedBox.tag)
      ]
    }

    // AES-GCM Decrypt
    AsyncFunction("decrypt") { (ciphertext: [UInt8], keyBytes: [UInt8], nonce: [UInt8], tag: [UInt8]) -> [UInt8] in
      let key = SymmetricKey(data: Data(keyBytes))
      let nonceData = Data(nonce)
      let ciphertextData = Data(ciphertext)
      let tagData = Data(tag)
      
      let sealedBox = try AES.GCM.SealedBox(
        nonce: AES.GCM.Nonce(data: nonceData),
        ciphertext: ciphertextData,
        tag: tagData
      )
      
      let decrypted = try AES.GCM.open(sealedBox, using: key)
      return Array(decrypted)
    }

    // Generate KeyPair (Ed25519)
    AsyncFunction("generateKeyPair") { () -> [String: Any] in
      let privateKey = Curve25519.Signing.PrivateKey()
      let publicKey = privateKey.publicKey
      
      return [
        "privateKey": Array(privateKey.rawRepresentation),
        "publicKey": Array(publicKey.rawRepresentation)
      ]
    }

    // Sign data with private key
    AsyncFunction("sign") { (data: [UInt8], privateKeyBytes: [UInt8]) -> [UInt8] in
      guard let privateKey = try? Curve25519.Signing.PrivateKey(rawRepresentation: Data(privateKeyBytes)) else {
        throw CryptoError.invalidKey
      }
      
      let signature = try privateKey.signature(for: Data(data))
      return Array(signature)
    }

    // Verify signature with public key
    AsyncFunction("verify") { (data: [UInt8], signature: [UInt8], publicKeyBytes: [UInt8]) -> Bool in
      guard let publicKey = try? Curve25519.Signing.PublicKey(rawRepresentation: Data(publicKeyBytes)) else {
        throw CryptoError.invalidKey
      }
      
      return publicKey.isValidSignature(Data(signature), for: Data(data))
    }

    // Generate random bytes
    AsyncFunction("generateRandomBytes") { (length: Int) -> [UInt8] in
      var bytes = [UInt8](repeating: 0, count: length)
      let status = SecRandomCopyBytes(kSecRandomDefault, bytes.count, &bytes)
      guard status == errSecSuccess else {
        throw CryptoError.failedToGenerateRandom
      }
      return bytes
    }
  }
}

enum CryptoError: Error {
  case failedToGenerateRandom
  case keyDerivationFailed
  case encryptionFailed
  case decryptionFailed
  case invalidKey
  case invalidSignature
}

File: crypto-native/android/src/main/java/expo/modules/cryptonative/CryptoNativeModule.kt
package expo.modules.cryptonative

import android.security.keystore.KeyGenParameterSpec
import android.security.keystore.KeyProperties
import expo.modules.kotlin.exception.CodedException
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import java.security.KeyPairGenerator
import java.security.MessageDigest
import java.security.Signature
import java.security.SecureRandom
import javax.crypto.Cipher
import javax.crypto.KeyGenerator
import javax.crypto.SecretKey
import javax.crypto.spec.GCMParameterSpec
import javax.crypto.spec.SecretKeySpec

class CryptoNativeModule : Module() {
  private val secureRandom = SecureRandom()

  override fun definition() = ModuleDefinition {
    Name("CryptoNative")

    // Generate a random salt
    AsyncFunction("generateSalt") { length: Int ->
      val salt = ByteArray(length)
      secureRandom.nextBytes(salt)
      salt.toList().map { it.toUByte().toInt() }
    }

    // Derive key using PBKDF2
    AsyncFunction("deriveKey") { password: String, salt: List<Int>, iterations: Int, keyLength: Int ->
      val saltBytes = salt.map { it.toByte() }.toByteArray()
      val factory = javax.crypto.SecretKeyFactory.getInstance("PBKDF2WithHmacSHA256")
      val spec = javax.crypto.spec.PBEKeySpec(password.toCharArray(), saltBytes, iterations, keyLength * 8)
      val key = factory.generateSecret(spec)
      key.encoded.toList().map { it.toUByte().toInt() }
    }

    // AES-GCM Encrypt
    AsyncFunction("encrypt") { data: List<Int>, keyBytes: List<Int> ->
      val plainText = data.map { it.toByte() }.toByteArray()
      val keyData = keyBytes.map { it.toByte() }.toByteArray()
      val secretKey = SecretKeySpec(keyData, "AES")
      
      val cipher = Cipher.getInstance("AES/GCM/NoPadding")
      cipher.init(Cipher.ENCRYPT_MODE, secretKey)
      
      val ciphertext = cipher.doFinal(plainText)
      val nonce = cipher.iv
      val tag = ciphertext.takeLast(16) // GCM tag is last 16 bytes
      
      mapOf(
        "ciphertext" to ciphertext.dropLast(16).map { it.toUByte().toInt() },
        "nonce" to nonce.toList().map { it.toUByte().toInt() },
        "tag" to tag.map { it.toUByte().toInt() }
      )
    }

    // AES-GCM Decrypt
    AsyncFunction("decrypt") { ciphertext: List<Int>, keyBytes: List<Int>, nonce: List<Int>, tag: List<Int> ->
      val keyData = keyBytes.map { it.toByte() }.toByteArray()
      val secretKey = SecretKeySpec(keyData, "AES")
      
      val nonceData = nonce.map { it.toByte() }.toByteArray()
      val ciphertextBytes = ciphertext.map { it.toByte() }.toByteArray()
      val tagBytes = tag.map { it.toByte() }.toByteArray()
      
      val fullCiphertext = ciphertextBytes + tagBytes
      
      val cipher = Cipher.getInstance("AES/GCM/NoPadding")
      val spec = GCMParameterSpec(128, nonceData)
      cipher.init(Cipher.DECRYPT_MODE, secretKey, spec)
      
      val decrypted = cipher.doFinal(fullCiphertext)
      decrypted.toList().map { it.toUByte().toInt() }
    }

    // Generate KeyPair (Ed25519/EdDSA)
    AsyncFunction("generateKeyPair") {
      // For Android, we'll use simple RSA keypair for compatibility
      // In production, consider BouncyCastle for Ed25519
      val keyPairGenerator = KeyPairGenerator.getInstance("RSA")
      keyPairGenerator.initialize(2048)
      val keyPair = keyPairGenerator.generateKeyPair()
      
      mapOf(
        "privateKey" to keyPair.private.encoded.toList().map { it.toUByte().toInt() },
        "publicKey" to keyPair.public.encoded.toList().map { it.toUByte().toInt() }
      )
    }

    // Sign data
    AsyncFunction("sign") { data: List<Int>, privateKeyBytes: List<Int> ->
      // Simplified signing - in production use proper Ed25519
      val signature = Signature.getInstance("SHA256withRSA")
      signature.initSign(java.security.KeyFactory.getInstance("RSA")
        .generatePrivate(java.security.spec.PKCS8EncodedKeySpec(privateKeyBytes.map { it.toByte() }.toByteArray())))
      signature.update(data.map { it.toByte() }.toByteArray())
      signature.sign().toList().map { it.toUByte().toInt() }
    }

    // Verify signature
    AsyncFunction("verify") { data: List<Int>, signatureBytes: List<Int>, publicKeyBytes: List<Int> ->
      try {
        val signature = Signature.getInstance("SHA256withRSA")
        signature.initVerify(java.security.KeyFactory.getInstance("RSA")
          .generatePublic(java.security.spec.X509EncodedKeySpec(publicKeyBytes.map { it.toByte() }.toByteArray())))
        signature.update(data.map { it.toByte() }.toByteArray())
        signature.verify(signatureBytes.map { it.toByte() }.toByteArray())
      } catch (e: Exception) {
        false
      }
    }

    // Generate random bytes
    AsyncFunction("generateRandomBytes") { length: Int ->
      val bytes = ByteArray(length)
      secureRandom.nextBytes(bytes)
      bytes.toList().map { it.toUByte().toInt() }
    }
  }
}
